src/receivers/fluent/receiver.rs

```diff
@@ -0,0 +1,852 @@
+                                        continue
+                                    }
+                                    if let Some(_) = &event_options.chunk {
+                                        error!("Fluent sender expects acks, not supported yet");
+                                    }
+                                }
+                                _ => {}
+                            }
+
+                            // This may exceed the batch size, but that's alright. Forwarded messages
+                            // may contain multiple entries, but we don't split them
+                            batch.push(message);
+
+                            // If we just hit or exceeded MAX_BATCH_SIZE, send the batch.
+                            if batch.size() >= MAX_BATCH_SIZE && pending_encode.is_none() {
+                                Self::send_batch(&mut batch, &mut pending_encode, &mut last_batch_send, Instant::now());
+                            }
+                        }
+                        Some(Err(e)) => {
+                            error!(
+                                connection_type = connection_type,
+                                "Error reading frame: {}", e
+                            );
+                            break;
+                        }
+                        None => {
+                            debug!(
+                                connection_type = connection_type,
+                                "Connection closed by peer"
+                            );
+                            break;
+                        }
+                    }
+                }
+
+                _ = self.cancel_token.cancelled() => {
+                    debug!(
+                        connection_type = connection_type,
+                        "Connection handler cancelled"
+                    );
+                    break;
+                }
+            }
+        }
+
+        let drain_timeout = tokio::time::Duration::from_secs(2); // TODO: make configurable
+        let drain_deadline = tokio::time::Instant::now()
+            .checked_add(drain_timeout)
+            .unwrap();
+
+        // Wait for remaining processing task to complete with timeout
+        while !batch.is_empty() || pending_encode.is_some() || pending_send.is_some() {
+            let now = Instant::now();
+
+            let time_left = drain_deadline.saturating_duration_since(now);
+            if time_left.is_zero() {
+                error!("Timed out draining fluent receiver");
+                break;
+            }
+
+            // Must empty the batch, otherwise drop into select and wait on
+            // the pending encode
+            if !batch.is_empty() && pending_encode.is_none() {
+                let curr_batch = batch.take();
+
+                pending_encode = Some(tokio::task::spawn_blocking(move || {
+                    Some(convert_to_otlp_logs(curr_batch))
```

Comment:
Do we need to consider partitioning of these batches into separate ResourceLogs and ScopeLogs where applicable? Particularly for the TCP connection endpoint? So for example if there is a tag that is "service.name" do we want to hoist that into a Resource attribute and partition these accordingly?